File: java/src/main/java/ch1_introduction_to_algorithms/BinarySearch.java
package ch1_introduction_to_algorithms;

public class BinarySearch {
    public static void main(String[] args) {
        int[] myList = {21, 45, 87, 93};

        System.out.println(binarySearch(myList, 93));
        System.out.println(binarySearch(myList, 16));
    }

    public static int binarySearch(int[] list, int target) {
        // Check if list is empty
        if (list.length == 0) {
            return -1;
        }

        int low = 0;
        int high = list.length - 1;

        while (low <= high) {
            int mid = (low + high) / 2;
            int guess = list[mid];

            if (guess == target) {
                return mid;
            } else if (guess > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: python/01_introduction_to_algorithms/binary_search.py
class BinarySearch:

    def binary_search(self, arr, item):
        # low and high keep track of which part of the list you'll search in.
        low = 0
        high = len(arr) - 1

        while low <= high:
            mid = (high + low) // 2 # integer (floor) division - rounds down
            guess = arr[mid]

            if guess == item:
                return mid
            elif guess > item:
                high = mid - 1
            else:
                low = mid + 1

        # If item does not exist
        return None

if __name__ == "__main__":
    bs = BinarySearch()
    my_list = [1, 3, 5, 7, 9]

    print(bs.binary_search(my_list, 3)) # => 1
    # 'None' means nil in Python. We use to indicate that the item wasn't found.
    print(bs.binary_search(my_list, -1)) # => None
